File: packages/twerk-slots/src/twerk_slots/cli/slot/free.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, NoReturn

import click

from twerk_core import get_console
from twerk_core.clinkr.dataclass_json import JsonSerializable
from twerk_core.clinkr.exit import ClinkrExit
from twerk_core.clinkr.operation import clinkr_operation
from twerk_slots.allocation import (
    DirtyWorktreeError,
    SlotAllocationError,
    SlotNotAssignedError,
    free_slot_assignment,
)
from twerk_slots.cli.slot.context import load_slots_context
from twerk_slots.cli.slot.selectors import (
    SelectorOk,
    SelectorResult,
    resolve_current,
    resolve_num,
    resolve_wt,
)
from twerk_slots.context import SlotsCliContext
from twerk_slots.pool_state import AssignmentMissing, PoolState
from twerk_slots.repo_context import NoRepoSentinel


@dataclass(frozen=True)
class SlotFreeRequest:
    num: Annotated[
        tuple[int, ...],
        click.Option(["-n", "--num"], type=click.INT, multiple=True),
    ] = ()
    wt: Annotated[
        tuple[str, ...],
        click.Option(["-w", "--wt"], type=click.STRING, multiple=True),
    ] = ()
    current: Annotated[bool, click.Option(["-c", "--current"], is_flag=True, default=False)] = False


@dataclass(frozen=True)
class FreedSlot:
    slot_name: str
    branch_name: str
    worktree_path: str


@dataclass(frozen=True)
class SlotFreeResult(JsonSerializable):
    freed: tuple[FreedSlot, ...]


def render_slot_free(result: SlotFreeResult) -> None:
    console = get_console()
    for entry in result.freed:
        console.print(
            f"[green]✓[/green] Freed [bold cyan]{entry.slot_name}[/bold cyan] "
            f"([yellow]{entry.branch_name}[/yellow])"
        )
        console.print(
            f"  Worktree kept at [dim]{entry.worktree_path}[/dim]; detached HEAD at trunk"
        )


def _resolve_targets(
    slots_ctx: SlotsCliContext,
    request: SlotFreeRequest,
    state: PoolState,
) -> tuple[tuple[str, ...], tuple[str, ...]]:
    """Resolve every selector to a slot name.

    Returns (resolved_in_order, errors). ``resolved_in_order`` preserves
    first-seen order across ``--num`` → ``--wt`` → ``--current`` and is
    deduped. ``errors`` collects every shape-level problem so the caller
    can surface them in a single combined message.
    """
    resolved: list[str] = []
    seen: set[str] = set()
    errors: list[str] = []

    def absorb(result: SelectorResult) -> None:
        if isinstance(result, SelectorOk):
            if result.slot_name not in seen:
                seen.add(result.slot_name)
                resolved.append(result.slot_name)
        else:
            errors.append(result.message)

    for value in request.num:
        absorb(resolve_num(value, state.pool_size))

    for value in request.wt:
        absorb(resolve_wt(value))

    if request.current:
        absorb(resolve_current(slots_ctx.repo.root))

    return tuple(resolved), tuple(errors)


def validate_assigned_and_clean(
    slots_ctx: SlotsCliContext,
    state: PoolState,
    targets: tuple[str, ...],
) -> tuple[str, ...]:
    """For each resolved slot, check it is assigned and the worktree is clean."""
    errors: list[str] = []
    for slot_name in targets:
        lookup = state.find_by_slot(slot_name)
        if isinstance(lookup, AssignmentMissing):
            errors.append(
                f"{slot_name} is not currently assigned. Run `slot list` to see the pool."
            )
            continue
        assignment = lookup.assignment
        if slots_ctx.git.has_uncommitted_changes(assignment.worktree_path):
            errors.append(
                f"{slot_name} has uncommitted changes at {assignment.worktree_path}. "
                f"Commit or stash before freeing."
            )
    return tuple(errors)


@clinkr_operation(
    name="free",
    help="Release one or more slot assignments; keep the worktree directories for reuse.",
    human_renderer=render_slot_free,
)
def run_free_slot(ctx: click.Context, request: SlotFreeRequest) -> ClinkrExit[SlotFreeResult]:
    slots_ctx = load_slots_context(ctx)
    if isinstance(slots_ctx, NoRepoSentinel):
        raise ClinkrExit.failure(error_type="not_in_repo", message=slots_ctx.message)

    if not slots_ctx.pool_state.exists():
        raise ClinkrExit.failure(
            error_type="pool_empty",
            message="No pool configured. Run `slot checkout` first.",
        )
    state = slots_ctx.pool_state.load()

    if not request.num and not request.wt and not request.current:
        raise ClinkrExit.failure(
            error_type="missing_slot_arg",
            message="Pass one of -n/--num, -w/--wt, or -c/--current to identify the slot.",
        )

    targets, shape_errors = _resolve_targets(slots_ctx, request, state)
    state_errors = validate_assigned_and_clean(slots_ctx, state, targets)
    all_errors = (*shape_errors, *state_errors)
    if all_errors:
        raise ClinkrExit.failure(
            error_type="invalid_slot_args",
            message="\n".join(all_errors),
        )

    freed: list[FreedSlot] = []
    for slot_name in targets:
        try:
            outcome = free_slot_assignment(slots_ctx, slot_name=slot_name)
        except SlotAllocationError as exc:
            partial_failure(freed, error_type="slot_allocation_error", message=str(exc))
        if isinstance(outcome, SlotNotAssignedError):
            partial_failure(
                freed,
                error_type="slot_not_assigned",
                message=f"{slot_name} is not currently assigned (state changed during free).",
            )
        if isinstance(outcome, DirtyWorktreeError):
            partial_failure(
                freed,
                error_type="dirty_worktree",
                message=(
                    f"{slot_name} has uncommitted changes at {outcome.worktree_path} "
                    f"(state changed during free)."
                ),
            )
        freed.append(
            FreedSlot(
                slot_name=outcome.slot_name,
                branch_name=outcome.branch_name,
                worktree_path=str(outcome.worktree_path),
            )
        )

    return ClinkrExit.ok(SlotFreeResult(freed=tuple(freed)))


def partial_failure(
    freed: list[FreedSlot],
    *,
    error_type: str,
    message: str,
) -> NoReturn:
    """Raise a `ClinkrExit.failure` that lists already-freed slots if any."""
    if not freed:
        raise ClinkrExit.failure(error_type=error_type, message=message)
    already = ", ".join(f.slot_name for f in freed)
    raise ClinkrExit.failure(
        error_type=error_type,
        message=f"{message} Already freed: {already}.",
    )

File: ts/packages/slot/src/lifecycle/free.ts
import type { SlotCliContext } from "../context.ts";
import { buildSlotInventory, findBySlot, type SlotRecord } from "../inventory.ts";
import { ensureSlotsMetadataDir } from "../repo-context.ts";
import { slotOperationMessage, type LifecycleResult } from "./common.ts";
import { cleanupErrorCount, executeReleaseCleanup, planReleaseCleanup, type SlotFreeCleanupAction, type SlotFreeCleanupResult } from "./release-cleanup.ts";
import { type FreedSlot, releaseAssignedSlotTarget, type ReleaseTarget } from "./release-target.ts";

export interface SlotFreePlan {
	targets: FreedSlot[];
	skipped: string[];
	cleanup: SlotFreeCleanupResult[];
	trunkBranch: string;
}

export interface SlotFreeOutcome {
	freed: FreedSlot[];
	would_free: FreedSlot[];
	cleanup: SlotFreeCleanupResult[];
	skipped: string[];
	dry_run: boolean;
	cancelled: boolean;
	cleanup_error_count: number;
}

export async function planFreeSlots(ctx: SlotCliContext, slotNames: readonly string[], options: { preflightErrors?: readonly string[] | undefined; skipped?: readonly string[] | undefined; cleanupActions?: readonly SlotFreeCleanupAction[] | undefined } = {}): Promise<LifecycleResult<SlotFreePlan>> {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	await ensureSlotsMetadataDir(ctx.repo, ctx.storage);
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	if (inventory.records.length === 0) return failure("pool_empty", "No managed slots configured. Run `slot init --size N` first.");
	const errors = [...options.preflightErrors ?? []];
	const targets: FreedSlot[] = [];
	for (const slotName of slotNames) {
		const record = findBySlot(inventory, slotName);
		const validation = await validateFreeTarget(ctx, record, slotName);
		if (validation.type === "error") errors.push(validation.message);
		else targets.push(validation.freed);
	}
	if (errors.length > 0) return failure("invalid_slot_args", errors.join("\n"));
	const trunkBranch = await ctx.git.getTrunkBranch();
	const cleanup = await planReleaseCleanup(ctx, targets, options.cleanupActions ?? []);
	return ok({ targets, skipped: [...options.skipped ?? []], cleanup: [...cleanup], trunkBranch });
}

export async function executeFreePlan(ctx: SlotCliContext, plan: SlotFreePlan, options: { cleanupActions?: readonly SlotFreeCleanupAction[] | undefined } = {}): Promise<LifecycleResult<SlotFreeOutcome>> {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	const freed: FreedSlot[] = [];
	for (const target of plan.targets) {
		const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
		const result = await releaseAssignedSlotTarget(ctx.git, inventory, { slotName: target.slot_name, branchName: target.branch_name }, plan.trunkBranch);
		if (result.type === "released") {
			freed.push(result.freed);
			continue;
		}
		const suffix = freed.length === 0 ? "" : ` Already freed: ${freed.map((slot) => slot.slot_name).join(", ")}.`;
		return failure(result.reason === "detach_failed" ? "slot_allocation_error" : result.reason, `${result.message}${suffix}`);
	}
	const cleanup = await executeReleaseCleanup(ctx, freed, options.cleanupActions ?? []);
	return ok({ freed, would_free: [], cleanup: [...cleanup], skipped: [...plan.skipped], dry_run: false, cancelled: false, cleanup_error_count: cleanupErrorCount(cleanup) });
}

export function dryRunFreeOutcome(plan: SlotFreePlan): SlotFreeOutcome {
	return { freed: [], would_free: [...plan.targets], cleanup: [...plan.cleanup], skipped: [...plan.skipped], dry_run: true, cancelled: false, cleanup_error_count: cleanupErrorCount(plan.cleanup) };
}

export function cancelledFreeOutcome(plan: SlotFreePlan): SlotFreeOutcome {
	return { freed: [], would_free: [...plan.targets], cleanup: [...plan.cleanup], skipped: [...plan.skipped], dry_run: false, cancelled: true, cleanup_error_count: 0 };
}

async function validateFreeTarget(ctx: SlotCliContext, record: SlotRecord | null, slotName: string): Promise<{ type: "ok"; freed: FreedSlot } | { type: "error"; message: string }> {
	if (record === null || record.branch === null) return { type: "error", message: `${slotName} is not currently assigned. Run \`slot list\` to see the pool.` };
	if (record.operation !== null) return { type: "error", message: slotOperationMessage(record, { action: "freeing" }) };
	if (await ctx.git.hasUncommittedChanges(record.path)) return { type: "error", message: `${record.slotName} has uncommitted changes at ${record.path}. Commit or stash before freeing.` };
	return { type: "ok", freed: { slot_name: record.slotName, branch_name: record.branch, worktree_path: record.path } };
}

function ok<T>(outcome: T): LifecycleResult<T> {
	return { type: "ok", outcome };
}

function failure<T>(errorType: string, message: string): LifecycleResult<T> {
	return { type: "failure", failure: { error_type: errorType, message } };
}
